rfd/0213-relay-a-lightweight-proxy.md

```diff
@@ -0,0 +1,64 @@
+
+Instead, agents configured to use a relay will connect to the relay group address to fetch the relay group discovery data (from any of the relays); this includes the relay group name, the agent connection count, and a list of the currently known relays for the group. They will then open reverse tunnel connections to different relays (up to the agent connection count) directly. This is different from the way that the regular reverse tunnel works, which requires agents to reach different proxies by opening new connections to the same public address to get randomly forwarded to different proxies behind a load balancer - connecting directly saves resources on the load balancer and the network, and it plays well with the intended usage scenario (agents and relays in the same subenvironment). Relays that are being gracefully terminated will report themselves as such to the connected agents (for both existing and new connections), and agents will not consider those relays for the purpose of filling the connection quota. Agents will include the list of relay host IDs that they're connected to in their resources' heartbeats (the same way that proxy IDs are advertised in heartbeats today if proxy peering is enabled). Relays that are terminating will not be included in agents' heartbeats, reverse tunnel connections to terminating relays will be closed a few minutes after the last tunneled connection, and the relays themselves will wait a few minutes after the last tunneled connection before exiting. With appropriate grace times for disconnections as part of the graceful shutdown of the Teleport relay agents it should be straightforward to scale
+
+The discovery data will be occasionally re-fetched as a safeguard against broken relays resulting in isolation of an agent, but it's expected that any change to the available relays as a result of rollouts and scaling will be quickly propagated to all relays through the usual event fanout, and will then be broadcast to all the connected agents (similarly to how the regular reverse tunnel works). It is not practical to have all relay heartbeats replicated to individual agents through their own event watchers, since heartbeat-like resources cached by large volumes of agents are a performance problem for the event fanout as well as a noticeable load on the network. The periodic fetching of the discovery data is also how agents will come to learn about changes to the agent connection count, if any.
+
+It should be possible for the relay reverse tunnels to support all services immediately, since no changes should be necessary for the connections tunneled from the Teleport control plane to the agents.
```

Comment:
Alternatively to this we could just refuse to start if any not-fully-supported services are enabled while `relay_server` is set.